core/state_processor.go

```diff
@@ -74,11 +77,29 @@ func applyTransaction(config *chain.Config, engine consensus.EngineReader, gp *G
 	}
 	*usedGas += result.UsedGas
 
+	var (
+		legacyReceipt     *types.LegacyReceipt
+		isBobaLegacyBlock bool
+	)
+	if evm.ChainConfig().IsBobaLegacyBlock(header.Number) {
+		isBobaLegacyBlock = true
+	}
+
+	if isBobaLegacyBlock && historicalRPCService != nil {
+		err = historicalRPCService.CallContext(context.Background(), &legacyReceipt, "eth_getTransactionReceipt", tx.Hash().String())
+		if err != nil {
+			return nil, nil, err
+		}
+		*usedGas = uint64(legacyReceipt.GasUsed)
+	} else if isBobaLegacyBlock {
+		// the legacy block must be handled by the historicalRPCService
```

Comment:
Nit, but I find a pattern of:

```
if cond1 && cond2 { } else if cond1 {return err... }
```

To be a bit confusing.  I'd personally write it as:

```
if cond1 { if !cond2 { return err... } }
```

Reply:
Fixed!
```
if isBobaLegacyBlock {
  if historicalRPCService != nil {
  } else {
  }
}
```